src/auto_renewal.cairo

```diff
@@ -0,0 +1,242 @@
+            self._is_renewing.write((caller, domain, limit_price), new_renew);
+
+            let prev_last_renewal = self.last_renewal.read((caller, domain));
+            // if we toggle the renewal on, we erase the previous renewal date
+            let new_last_renewal = prev_last_renewal * prev_renew;
+            self.last_renewal.write((caller, domain), new_last_renewal);
+
+            self
+                .emit(
+                    Event::toggled_renewal(
+                        toggled_renewal {
+                            domain,
+                            renewer: caller,
+                            limit_price,
+                            is_renewing: new_renew,
+                            last_renewal: new_last_renewal
+                        }
+                    )
+                )
+        }
+
+        fn renew(
+            ref self: ContractState,
+            root_domain: felt252,
+            renewer: ContractAddress,
+            limit_price: u256,
+        ) {
+            self._renew(root_domain, renewer, limit_price);
+        }
+
+        fn batch_renew(
+            ref self: ContractState,
+            domain: array::Array::<felt252>,
+            renewer: array::Array::<starknet::ContractAddress>,
+            limit_price: array::Array::<u256>,
+        ) {
+            assert(domain.len() == renewer.len(), 'Domain & renewer mismatch len');
+            assert(domain.len() == limit_price.len(), 'Domain & price mismatch len');
+
+            let mut domain = domain;
+            let mut renewer = renewer;
+            let mut limit_price = limit_price;
+
+            loop {
+                if domain.len() == 0 {
+                    break;
+                }
+                let _domain = domain.pop_front().expect('pop_front error');
+                let _renewer = renewer.pop_front().expect('pop_front error');
+                let _limit_price = limit_price.pop_front().expect('pop_front error');
+                self._renew(_domain, _renewer, _limit_price);
+            }
+        }
+    }
+
+    //
+    // Internals
+    //
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn _renew(
+            ref self: ContractState,
+            root_domain: felt252,
+            renewer: ContractAddress,
+            limit_price: u256,
+        ) {
+            let naming = self.naming_contract.read();
+            let can_renew = self._is_renewing.read((renewer, root_domain, limit_price));
+            assert(can_renew == 1, 'Renewal not toggled for domain');
+
+            // Check domain has not been renewed yet this year
+            let block_timestamp: felt252 = get_block_timestamp().into();
+            let last_renewed = self.last_renewal.read((renewer, root_domain));
+            let elapsed: u256 = u256_from_felt252(block_timestamp - last_renewed);
+            let max = u256_from_felt252(86400 * 364);
+            assert(elapsed > max, 'Domain already renewed');
```

Comment:
Please use a constant for max if possible. I fear that u256_from_felt252 would be computed on runtime and not on compilation. If using a constant is not possible and the u256 split from felt256 is actually done on runtime, please hardcode a u256 with this in low value and 0 in high value.